File: scripts/solar/solar-common/actor/state/SolarActorBuffState.lua
local ____lualib = require("lualib_bundle")
local __TS__Class = ____lualib.__TS__Class
local __TS__New = ____lualib.__TS__New
local __TS__SourceMapTraceBack = ____lualib.__TS__SourceMapTraceBack
__TS__SourceMapTraceBack(debug.getinfo(1).short_src, {["7"] = 5,["8"] = 5,["9"] = 6,["10"] = 6,["11"] = 7,["12"] = 7,["13"] = 8,["14"] = 8,["15"] = 9,["16"] = 9,["17"] = 10,["18"] = 10,["19"] = 12,["20"] = 12,["21"] = 13,["22"] = 13,["23"] = 15,["24"] = 15,["25"] = 16,["26"] = 16,["27"] = 17,["28"] = 17,["29"] = 20,["30"] = 20,["31"] = 20,["33"] = 43,["34"] = 44,["37"] = 49,["38"] = 50,["41"] = 53,["42"] = 54,["43"] = 55,["44"] = 53,["45"] = 57,["46"] = 58,["47"] = 59,["48"] = 57,["49"] = 61,["50"] = 62,["51"] = 63,["52"] = 61,["53"] = 65,["54"] = 66,["55"] = 67,["56"] = 65,["57"] = 71,["58"] = 72,["61"] = 75,["62"] = 75,["63"] = 75,["64"] = 77,["65"] = 78,["67"] = 75,["68"] = 75,["69"] = 71,["70"] = 85,["71"] = 86,["72"] = 86,["73"] = 86,["74"] = 86,["75"] = 86,["76"] = 86,["77"] = 86,["78"] = 86,["79"] = 88,["80"] = 89,["81"] = 90,["82"] = 93,["83"] = 95,["84"] = 96,["85"] = 99,["86"] = 100,["87"] = 101,["89"] = 42,["90"] = 106,["91"] = 107,["92"] = 108,["93"] = 109,["94"] = 110,["95"] = 111,["96"] = 112,["98"] = 114,["99"] = 116,["100"] = 117,["102"] = 120,["103"] = 121,["104"] = 122,["106"] = 106,["107"] = 126,["108"] = 127,["109"] = 128,["110"] = 126,["111"] = 134,["112"] = 135,["113"] = 136,["114"] = 137,["117"] = 142,["118"] = 143,["119"] = 144,["122"] = 148,["123"] = 149,["124"] = 150,["125"] = 151,["127"] = 152,["128"] = 153,["129"] = 154,["130"] = 155,["131"] = 156,["132"] = 160,["134"] = 162,["135"] = 163,["136"] = 164,["137"] = 165,["138"] = 166,["139"] = 168,["140"] = 169,["144"] = 172,["145"] = 173,["146"] = 172,["147"] = 134,["148"] = 182,["149"] = 183,["150"] = 184,["151"] = 183,["152"] = 182,["153"] = 23,["154"] = 30,["155"] = 30,["156"] = 30,["157"] = 31,["158"] = 32,["159"] = 33,["160"] = 34,["161"] = 35,["162"] = 36,["163"] = 30,["164"] = 30,["165"] = 38,["166"] = 39,["167"] = 40});
local ____exports = {}
local ____trigger = require("solar.solar-common.w3ts.handles.trigger")
local Trigger = ____trigger.Trigger
local ____ObjectPool = require("solar.solar-common.tool.ObjectPool")
local ObjectPool = ____ObjectPool.default
local ____frame = require("solar.solar-common.w3ts.handles.frame")
local Frame = ____frame.Frame
local ____SelectUtil = require("solar.solar-common.util.unit.SelectUtil")
local SelectUtil = ____SelectUtil.default
local ____ColorStr = require("solar.solar-common.constant.ColorStr")
local ColorStr = ____ColorStr.default
local ____SingletonUtil = require("solar.solar-common.util.lang.SingletonUtil")
local SingletonUtil = ____SingletonUtil.default
local ____SolarBuffState = require("solar.solar-common.attribute.SolarBuffState")
local SolarBuffState = ____SolarBuffState.default
local ____BuffUtil = require("solar.solar-common.util.system.BuffUtil")
local BuffUtil = ____BuffUtil.default
local ____ActorBuffUtil = require("solar.solar-common.actor.util.ActorBuffUtil")
local ActorBuffUtil = ____ActorBuffUtil.default
local ____AttributeUtil = require("solar.solar-common.util.system.AttributeUtil")
local AttributeUtil = ____AttributeUtil.default
local ____BaseUtil = require("solar.solar-common.util.BaseUtil")
local BaseUtil = ____BaseUtil.default
____exports.default = __TS__Class()
local SolarActorBuffState = ____exports.default
SolarActorBuffState.name = "SolarActorBuffState"
function SolarActorBuffState.prototype.____constructor(self)
    if SingletonUtil:notFirstTime(____exports.default) then
        print("不能重复new SolarActorBuffState()")
        return
    end
    if SingletonUtil:notFirstTime(SolarBuffState) then
        print("演员buff系统不能与旧buff系统同时使用!")
        return
    end
    BuffUtil.registerBuffType = function(____, buffType)
        log.errorWithTraceBack("演员buff系统不能与旧buff系统同时使用！")
        return nil
    end
    BuffUtil.getBuffType = function(____, buffTypeId)
        log.errorWithTraceBack("演员buff系统不能与旧buff系统同时使用！")
        return nil
    end
    BuffUtil.forAllBuffTypes = function(____, callback, buffTypeClass)
        log.errorWithTraceBack("演员buff系统不能与旧buff系统同时使用！")
        return nil
    end
    BuffUtil.addBuff = function(____, unit, buffTypeId, caster, initBuff)
        log.errorWithTraceBack("演员buff系统不能与旧buff系统同时使用！")
        return nil
    end
    ActorBuffUtil:addAnyActorBuffLevelChangeListener(function(____, buff, delta)
        if buff.attribute == nil or buff.actorType.attribute == nil then
            return
        end
        BaseUtil.runLater(
            0.1,
            function()
                if AttributeUtil:isEquals(buff.attribute, buff.actorType.attribute) then
                    buff.attribute = AttributeUtil:multiply(buff.actorType.attribute, buff.level)
                end
            end
        )
    end)
    if ____exports.default.config.showSolarActorBuffUI then
        ____exports.default.ubertipFrameBD = __TS__New(
            Frame,
            "BACKDROP",
            nil,
            nil,
            "_sl_border_backdrop",
            0
        )
        ____exports.default.ubertipFrameText = Frame:createTEXT()
        ____exports.default.ubertipFrameText:setAbsPoint(FRAMEPOINT_BOTTOMRIGHT, 0.78, 0.17)
        ____exports.default.ubertipFrameText:setSize(0.16, 0)
        ____exports.default.ubertipFrameBD:setPoints(____exports.default.ubertipFrameText.current, 0.01, 0.01)
        ____exports.default.ubertipFrameBD.visible = false
        ____exports.default.ubertipFrameText.visible = false
        local trigger = __TS__New(Trigger)
        trigger:registerTimerEvent(0.1, true)
        trigger:addAction(self.refreshSolarBuffUI)
    end
end
function SolarActorBuffState.showUbertip()
    local frameH = DzGetTriggerUIEventFrame()
    local actorBuff = ____exports.default.frameBuffMap["f" .. tostring(frameH)]
    if actorBuff then
        local buffTips = (ColorStr.yellow .. (actorBuff:getName() or "")) .. "|n|n"
        if actorBuff.level and actorBuff.level ~= 1 then
            buffTips = (((buffTips .. ColorStr.green) .. "等级: ") .. tostring(actorBuff.level or "1")) .. "|n|n"
        end
        buffTips = (buffTips .. ColorStr.white) .. (actorBuff:getDescribe() or "")
        if actorBuff:get("dur") and actorBuff:get("dur") > 3 then
            buffTips = ((buffTips .. "|n剩余时间:") .. tostring(math.ceil(actorBuff:getRemainingTime()))) .. "秒"
        end
        ____exports.default.ubertipFrameText:setText(buffTips)
        ____exports.default.ubertipFrameBD.visible = true
        ____exports.default.ubertipFrameText.visible = true
    end
end
function SolarActorBuffState.hideUbertip()
    ____exports.default.ubertipFrameBD.visible = false
    ____exports.default.ubertipFrameText.visible = false
end
function SolarActorBuffState.prototype.refreshSolarBuffUI()
    local realSelectUnit = SelectUtil.getRealSelectUnit()
    if not IsHandle(realSelectUnit) then
        ____exports.default:showBuffsUI(false)
        return
    end
    local unitBuffs = ActorBuffUtil:getUnitActorBuffs(realSelectUnit)
    if not unitBuffs then
        ____exports.default:showBuffsUI(false)
        return
    end
    local objectPool = ____exports.default.objectPool
    objectPool:setAllIdleStatus(true)
    local index = 0
    for actorTypeId in pairs(unitBuffs) do
        do
            local actorBuff = unitBuffs[actorTypeId]
            local uiShowType = actorBuff:get("uiShowType")
            if uiShowType ~= nil and uiShowType ~= "单位状态栏" then
                goto __continue22
            elseif actorBuff:get("class") and actorBuff:get("class") ~= "基础" and actorBuff:get("class") ~= "演员buff" and actorBuff:get("class") ~= "羁绊" and uiShowType ~= "单位状态栏" then
                goto __continue22
            end
            local frames = objectPool:borrowObject()
            local frame = frames[1]
            frame.visible = true
            frames[2]:setTexture(actorBuff:get("icon", "ReplaceableTextures\\CommandButtons\\BTNTemp.blp"))
            frame:setAbsPoint(FRAMEPOINT_BOTTOMLEFT, 0.37 + index * 0.016, 0.007)
            ____exports.default.frameBuffMap["f" .. tostring(frame.current)] = actorBuff
            index = index + 1
        end
        ::__continue22::
    end
    objectPool:forDirtyObjects(function(____, frames)
        frames[1].visible = false
    end)
end
function SolarActorBuffState.showBuffsUI(self, show)
    ____exports.default.objectPool:forActiveObjects(function(____, frames)
        frames[1].visible = show
    end)
end
SolarActorBuffState.config = {showSolarActorBuffUI = true}
SolarActorBuffState.objectPool = __TS__New(
    ObjectPool,
    function()
        local frame = Frame:createBUTTON()
        local backgroundImage = frame:addBackgroundImage("")
        frame:setOnMouseEnter(____exports.default.showUbertip)
        frame:setOnMouseLeave(____exports.default.hideUbertip)
        frame:setSize(0.015, 0.015)
        return {frame, backgroundImage}
    end
)
SolarActorBuffState.frameBuffMap = {}
SolarActorBuffState.ubertipFrameBD = nil
SolarActorBuffState.ubertipFrameText = nil
return ____exports

File: src/solar/solar-common/actor/state/SolarActorBuffState.ts
/**
 * 太阳Buff系统 调度中心
 *
 */
import {Trigger} from "@/w3ts/handles/trigger";
import ObjectPool from "@/ObjectPool";
import {Frame} from "@/w3ts/handles/frame";
import SelectUtil from "@/util/unit/SelectUtil";
import ColorStr from "@/ColorStr";
import SingletonUtil from "@/SingletonUtil";
import ActorBuff from "@/ActorBuff";
import SolarBuffState from "@/SolarBuffState";
import BuffUtil from "@/BuffUtil";
import Buff from "@/Buff";
import ActorBuffUtil from "@/ActorBuffUtil";
import AttributeUtil from "@/AttributeUtil";
import BaseUtil from "@/BaseUtil";


export default class SolarActorBuffState {
    static config: {
        showSolarActorBuffUI //显示太阳buff UI
    } = {
        showSolarActorBuffUI: true
    }

    /**
     * 模拟buff的 ui显示
     */
    static objectPool = new ObjectPool<Frame[]>(() => {
        let frame = Frame.createBUTTON();
        let backgroundImage = frame.addBackgroundImage("");
        frame.setOnMouseEnter(SolarActorBuffState.showUbertip)
        frame.setOnMouseLeave(SolarActorBuffState.hideUbertip)
        frame.setSize(0.015, 0.015)
        return [frame, backgroundImage];
    });
    static frameBuffMap: { [frameId: string]: ActorBuff } = {}
    static ubertipFrameBD: Frame = null;
    static ubertipFrameText: Frame = null;

    constructor() {
        if (SingletonUtil.notFirstTime(SolarActorBuffState)) {
            print("不能重复new SolarActorBuffState()")
            return;
        }

        //冲突检测
        if (SingletonUtil.notFirstTime(SolarBuffState)) {
            print("演员buff系统不能与旧buff系统同时使用!")
            return;
        }
        BuffUtil.registerBuffType = (buffType: AppBuffType) => {
            log.errorWithTraceBack("演员buff系统不能与旧buff系统同时使用！")
            return null;
        }
        BuffUtil.getBuffType = (buffTypeId: string) => {
            log.errorWithTraceBack("演员buff系统不能与旧buff系统同时使用！")
            return null;
        }
        BuffUtil.forAllBuffTypes = (callback: (buffType: AppBuffType) => void, buffTypeClass?: string) => {
            log.errorWithTraceBack("演员buff系统不能与旧buff系统同时使用！")
            return null;
        }
        BuffUtil.addBuff = (unit: unit, buffTypeId: string, caster?: unit, initBuff?: (buff: Buff) => void) => {
            log.errorWithTraceBack("演员buff系统不能与旧buff系统同时使用！")
            return null;
        }

        //注册任意buff层数改变事件 修改对应的属性附加值
        ActorBuffUtil.addAnyActorBuffLevelChangeListener((buff, delta) => {
            if (buff.attribute == null || buff.actorType.attribute == null) {
                return
            }
            BaseUtil.runLater(0.1, () => {
                //如果没有手动指定属性的话 则自动根据等级修改属性
                if (AttributeUtil.isEquals(buff.attribute, buff.actorType.attribute)) {
                    buff.attribute = AttributeUtil.multiply(buff.actorType.attribute, buff.level)
                }
            });
        });


        // buff ui显示
        if (SolarActorBuffState.config.showSolarActorBuffUI) {
            SolarActorBuffState.ubertipFrameBD = new Frame("BACKDROP", null, null, "_sl_border_backdrop", 0)
            //
            SolarActorBuffState.ubertipFrameText = Frame.createTEXT()
            SolarActorBuffState.ubertipFrameText.setAbsPoint(FRAMEPOINT_BOTTOMRIGHT, 0.78, 0.17)
            SolarActorBuffState.ubertipFrameText.setSize(0.16, 0)


            SolarActorBuffState.ubertipFrameBD.setPoints(SolarActorBuffState.ubertipFrameText.current, 0.01, 0.01);
            // SolarBuffState.ubertipFrameBD.setTexture("UI\\Glues\\BattleNet\\BattleNetTeamLevelBar\\Loading-BarBackground.blp")
            SolarActorBuffState.ubertipFrameBD.visible = false;
            SolarActorBuffState.ubertipFrameText.visible = false;

            //
            let trigger = new Trigger();
            trigger.registerTimerEvent(0.1, true);
            trigger.addAction(this.refreshSolarBuffUI)
        }

    }

    static showUbertip(this: void) {
        let frameH = DzGetTriggerUIEventFrame();
        let actorBuff = SolarActorBuffState.frameBuffMap["f" + frameH];
        if (actorBuff) {
            let buffTips = ColorStr.yellow + (actorBuff.getName() ?? "") + "|n|n";
            if (actorBuff.level && actorBuff.level != 1) {
                buffTips = buffTips + ColorStr.green + "等级: " + (actorBuff.level ?? "1") + "|n|n"
            }
            buffTips = buffTips + ColorStr.white + (actorBuff.getDescribe() ?? "")
            //超过3秒的才显示 部分光环buff可能就是3秒 然后循环添加这个buff 这种就不会显示剩余时间
            if (actorBuff.get("dur") && actorBuff.get("dur") > 3) {
                buffTips = buffTips + "|n剩余时间:" + Math.ceil(actorBuff.getRemainingTime()) + "秒"
            }

            SolarActorBuffState.ubertipFrameText.setText(buffTips)
            SolarActorBuffState.ubertipFrameBD.visible = true;
            SolarActorBuffState.ubertipFrameText.visible = true;
        }
    }

    static hideUbertip(this: void) {
        SolarActorBuffState.ubertipFrameBD.visible = false;
        SolarActorBuffState.ubertipFrameText.visible = false;
    }

    /**
     * 更新太阳buff
     */
    refreshSolarBuffUI(this: void) {
        let realSelectUnit = SelectUtil.getRealSelectUnit();
        if (!IsHandle(realSelectUnit)) {
            SolarActorBuffState.showBuffsUI(false);
            return;
        }


        let unitBuffs = ActorBuffUtil.getUnitActorBuffs(realSelectUnit);
        if (!unitBuffs) {
            SolarActorBuffState.showBuffsUI(false);
            return;
        }
        //
        let objectPool = SolarActorBuffState.objectPool;
        objectPool.setAllIdleStatus(true);//设置对象池所有 ui为 空闲可使用状态
        let index = 0;
        for (let actorTypeId in unitBuffs) {
            let actorBuff = unitBuffs[actorTypeId];
            let uiShowType = actorBuff.get("uiShowType");
            if (uiShowType != null && uiShowType != "单位状态栏") {
                continue
            } else if (actorBuff.get("class") && actorBuff.get("class") != "基础" &&
                actorBuff.get("class") != "演员buff" &&
                actorBuff.get("class") != "羁绊" &&
                uiShowType != "单位状态栏") {
                continue;//不是基础类型的其他类型则不在这里显示
            }
            let frames = objectPool.borrowObject();
            let frame = frames[0];
            frame.visible = true
            frames[1].setTexture(actorBuff.get("icon", "ReplaceableTextures\\CommandButtons\\BTNTemp.blp"))
            frame.setAbsPoint(FRAMEPOINT_BOTTOMLEFT, 0.37 + (index * 0.016), 0.007)
            //绑定ui跟 actorBuff
            SolarActorBuffState.frameBuffMap["f" + frame.current] = actorBuff;
            index++;
        }

        objectPool.forDirtyObjects(frames => {
            frames[0].visible = false;
        })


    }

    /**
     * UI显示
     */
    static showBuffsUI(show: boolean) {
        SolarActorBuffState.objectPool.forActiveObjects(frames => {
            frames[0].visible = show;
        });
    }


}